Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

4ntoine opened the pull request:
Now elemhide every blocked visible resource.

@xxxz @shoniko

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Now elemhide every blocked visible resource.

4ntoine commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,95 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+ if (elements.length > 0) {\n\
+   //console.log('found ' + elements.length + ' elemhide blocked elements for selector ' + selector);\n\
+   for (var e=0; e<elements.length; e++) {\n\
+     hideElement(elements[e]);\n\
```
i call only `hideElement` and in original sources it's not used sometimes:
https://github.com/adblockplus/adblockpluschrome/blob/master/include.preload.js#L224

4ntoine commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,95 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+    LOG(WARNING) << "Adblock: frameTreeNodeId=" << frameTreeNodeId;
+
+    std::string js = generateBlockedElemhideJavaScript(url);
+    //LOG(WARNING) << "Adblock: generated elemhide js for blocked url\n" << js;
```
commented line left for debugging purpose.
probably makes sense to delete the line

4ntoine commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,95 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+ //console.log('blocked element selector: ' + selector);\n\
+ var elements = document.querySelectorAll(selector);\n\
+ if (elements.length > 0) {\n\
+   //console.log('found ' + elements.length + ' elemhide blocked elements for selector ' + selector);\n\
```
commented line left for debugging purpose.
probably makes sense to delete the line

shoniko commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,95 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
   return rv;
 }
 
+std::string generateBlockedElemhideJavaScript(std::string url) {
```
Is there any way we can store JS like this in a resource instead? We could then potentially sync it from adblockpluschrome directly when building, but also it would be just easier to maintain.

xxxz replied:
agree. I am also fine if we land it with the creation of an issue to use it from a resource or even to generate a sequence of bytes as in libadblockplus.

xxxz commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,95 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
   return rv;
 }
 
+std::string generateBlockedElemhideJavaScript(std::string url) {
```
generally we pass `const std::string& url`

xxxz commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,95 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+  // basically copy-pasting JS from
+  // https://github.com/adblockplus/adblockpluschrome/blob/master/include.preload.js#L546
+  ss <<
+"function hideElement(element)\n\
```
It can be easier with C++ raw strings. Could you please change it, if there is no easy way with resources yet.

4ntoine replied:
i'm investigating how we can bundle js files from `core` module. Though it still sounds not very useful because js files will have to be modified for AdblockChromium specially: `required` keyword will not work in AdblockChromium.

xxxz replied:
Yeah, we can discuss it separately. Meanwhile, it seems we can just return a raw string.
BTW, it would be better to use streams to set the URL or even just replace some template entry in the string. E.g. https://github.com/adblockplus/adblockplusie/blob/05f8b81085d9da9ae03d7e5ddd51ba47c491d01d/src/shared/Utils.cpp#L174 (pay attention that it's ABP for IE project).

xxxz commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,95 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+ var elements = document.querySelectorAll(selector);\n\
+ if (elements.length > 0) {\n\
+   //console.log('found ' + elements.length + ' elemhide blocked elements for selector ' + selector);\n\
+   for (var e=0; e<elements.length; e++) {\n\
```
What about using of `for..of` loops or even `forEach`?

4ntoine replied:
is it better in any way?
BTW in `adblock_bridge.cc` we iterate using index too.

xxxz replied:
It's mainly consistent with other JS code (in core, e.g.). Additionally `document.querySelectorAll(selector).forEach(hideElement)` is much less code, what is also better.

4ntoine replied:
it's indeed more concise unless we want to output array size and have some additional checks (not just call some other function). Now we need to check each element `src`

xxxz replied:
still, I would propose to change it to `for..of`

4ntoine replied:
i still don't see any point

xxxz replied:
It will be less code, without the integer index variable and consistent with the other JS code. Additionally it's better to use `i` instead of `e`.

4ntoine replied:
it will be inconsistent with existing code. It could be done in separate "beautify" task

xxxz replied:
OK, however it conflicts with https://adblockplus.org/coding-style#javascript

4ntoine replied:
current code style assumes you have full debugging possibility. And if you "debug" using android logcat one have to have code that outputs debug values and does not require changing of iteration loop kind from "modern" if you need to see one value (array length).

abby-sergz replied:
But one can still see the array length

4ntoine replied:
if we chain it like you suggest `document.querySelectorAll(selector).forEach(hideElement)` we can't do it

xxxz replied:
I would suggest
```
let elements = document.querySelectorAll(selector);
if (elements.length > 0) {
  console.debug(`'found ${elements.length} elemhide blocked elements for selector '${selector}'`);
  for (let element of elements) {
    if (element.src == url)
      hideElement(element);
  }
}
```
BTW, we use `let` instead of `var`. Additionally, I would try to use `console.debug` and [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) and rather somehow introduce the variable `url`.

xxxz commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -402,10 +494,14 @@ int ChromeNetworkDelegate::OnBeforeStartTransaction(
         AdblockPlus::FilterEngine::ContentType adblock_content_type;
 
+        // some some resource types we need to do element hiding as they take empty screen space
+        bool elemHide;
```
could you please initialize it to `false`?

4ntoine replied:
sure, though i expect it to be `false` by default.

xxxz replied:
Just for reference, https://en.cppreference.com/w/cpp/language/default_initialization

> otherwise, nothing is done: the objects with automatic storage duration (and their subobjects) are initialized to indeterminate values.

and examples below

> int x;               // OK: the value of x is indeterminate

4ntoine replied:
thanks for the link - that differs from java.

hfiguiere replied:
the compiler should have thrown a warning (unless it has been disabled)

4ntoine replied:
fixed with explicitly setting of default value

xxxz commented:
BTW, I just hope that no object accessed by local variables disappears while our posted to UI thread callback is being executed.

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Fixed for relative URLs (simple-adblock.com).

hfiguiere commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,98 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
 }
 
+std::string generateBlockedElemhideJavaScript(std::string url, std::string relative_url) {
+  std::stringstream ss;
```
you could just return the string literal here.

4ntoine replied:
later `std::string` (`base::string16`) is required to call `ExecuteJavaScript..`. Is there any point to return literal and then convert to `base::string16`?

hfiguiere replied:
My point it that usimng `std::stringstream` here is not optimal, whether you return a `std::string` or the result of `base::UTF8ToUTF16()`.

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Set initial value for `elemHide`, minor typo fix.

xxxz commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +330,98 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+
+    if (frameHost) {
+      // elemhide resource by elemehide rules
+      frameHost->ExecuteJavaScriptInIsolatedWorld(
```
BTW, sorry I didn't look into the code of chromium. Does it execute the script completely isolated from the page or it can be clashed with the script of the page?

4ntoine replied:
it runs in isolated world (the same we use for element hiding for element hiding rules)

xxxz replied:
Is the "world" new for each invocation or it's shared among several invocations?

4ntoine replied:
does it make sense?

i believe it's shared between invocations, but we can easily check it by calling 2 scripts with the same isolated world id in a row and let's say 1st creates variable and 2nd reads it.

https://developer.chrome.com/extensions/content_scripts#isolated_world

xxxz replied:
Could you try it? If it's this way then we should re-use functions or wrap each injected JS by anonymous JS functions.

4ntoine replied:
it looks like not very significant improvement and could be done in separate task. i will be able to check it tomorrow (compiling for #41 at the moment)

xxxz replied:
If there are at least two blocked requests then starting from the second one that JS will not work because one may not declare variables several times in JS using `let` and `const`.

4ntoine replied:
how can you explain that there are 3 blocked images on 50statesclassifieds.com and all of them are elemhidden? see screenshots at https://gitlab.com/eyeo/adblockplus/chromium/issues/12

xxxz replied:
Can it be that for each frame there is a separate "world"?

4ntoine replied:
It can be. I've just checked on 50statesclassifieds.com each image is in separate iframe.
So you're saying that JS for the 2nd blocked resource (and all next) in iframe will fail because in the JS for the 1st we have same consts (`let`), right?

xxxz replied:
Yes, we should ensure that it works in the same frame because one can declare a variable only once with `let`.

4ntoine replied:
we could test it with testpages, but it's blocked by https://gitlab.com/eyeo/adblockplus/chromium/issues/39.

I will try to reproduce with some other URL with 2 blocked resources at the same frame

4ntoine replied:
@xxxz  it seems you're right. testing on theverge.com:
> 08-29 17:11:29.113 22920-22920 I/chromium: [INFO:CONSOLE(1)] "Uncaught SyntaxError: Identifier 'selector' has already been declared", source: https://cdn.krxd.net/partnerjs/xdi/proxy.3d2100fd7107262ecb55ce6847f01fa5.html#!kxcid=JImcjrMY&kxt=https%3A%2F%2Fcdn.krxd.net&kxcl=cdn&kxp= (1)

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Moved JS from core to resources.

4ntoine commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +332,76 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+  std::stringstream ss;
+
+  // basically copy-pasted JS and saved to resources from
+  // https://github.com/adblockplus/adblockpluschrome/blob/master/include.preload.js#L546
```
i will create separate task for `core` to extract files to be used both in `core` and on mobile and separate (dependent) task for `adblockchromium` to checkout `core` and reuse js from it (without hardcoding to resources). so we will be able to call `gclient sync` and fetch `core` files and reuse them here.

4ntoine commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +332,76 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+    .as_string();
+  ss << elemhide_js;
+
+  ss <<
```
i'm feeling strange about this part.
We can use Grit **messages** and insert values on-the-fly, but we will need to format JS as XML and it's undesired. Another option is to store it as file (like JS from `core` above) with our own tokens like `%url%` and replace tokens to actual values - looks a bit not perfect too.

Leave as-is for now.

4ntoine commented on third_party/libadblockplus/grit/js_resources.grd:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
copy-paste from `core` - to be fetched from core and reused (see task description above).

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

 Removed comments from .grd file.

4ntoine commented on third_party/libadblockplus/grit/resources/elemhide.js:
```diff
@@ -0,0 +1,26 @@
+function hideElement(element)
```
copy-paste from core - to be fetched from core and reused (see task description above).

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Using string instead of stringstream.

xxxz commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +332,76 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
   return rv;
 }
 
+std::string generateBlockedElemhideJavaScript(std::string url, std::string relative_url) {
```
the arguments should be `const std::string&`

4ntoine replied:
ACK

xxxz commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +332,76 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
+"\n\
+ var selector = \"[src$='" + relative_url + "'], [srcset$='" + relative_url + "']\";\n\
+ //console.log('blocked element selector: ' + selector);\n\
+ var elements = document.querySelectorAll(selector);\n\
```
So, what about using of raw strings? E.g. https://gitlab.com/eyeo/adblockplus/libadblockplus/blob/master/test/FileSystemJsObject.cpp#L282, https://en.cppreference.com/w/cpp/language/string_literal

4ntoine replied:
what about having it in resources and replacing some token with actual values ? As for me it's better than hardcoding of JS

xxxz commented on chrome/browser/net/chrome_network_delegate.cc:
```diff
@@ -327,6 +332,76 @@ int ChromeNetworkDelegate::OnBeforeURLRequest(
   return rv;
 }
 
+std::string generateBlockedElemhideJavaScript(std::string url, std::string relative_url) {
```
`url` argument is not used

4ntoine replied:
it is used

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

 One more minor change.

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Moved one more JS to resources, clean-up.

4ntoine commented on chrome/browser/browser_resources.grd:
```diff
@@ -656,6 +656,8 @@
         <include name="IDR_TAB_RANKER_EXAMPLE_PREPROCESSOR_CONFIG_PB" file="resource_coordinator\tab_ranker\example_preprocessor_config.pb" type="BINDATA" />
       </if>
+      <include name="IDR_ADBLOCK_ELEMHIDE_JS" file="../../third_party/libadblockplus/grit/resources/elemhide.js" type="BINDATA" />
+      <include name="IDR_ADBLOCK_ELEMHIDE_FOR_SELECTOR_JS" file="../../third_party/libadblockplus/grit/resources/elemhide_for_selector.jst" type="BINDATA" />
```
.jst stands for "JavaScriptTemplate". if we call it just ".js" "js_minify.py" script will be called and fails on packing of js because of `{{..}}` tokens used. Probably we can choose another replace token like `/*url*/` to be js-valid

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Moved to modern JS.

xxxz commented on third_party/libadblockplus/grit/resources/elemhide_for_selector.jst:
```diff
@@ -0,0 +1,9 @@
+// {{filename_with_query}} and {{url}} will be replaced with actual variable values
```
I also thought to move that snippet to a function, perhaps even to place it into *elemhide.js* and here left just a single line which is calling that function and passes the values.

4ntoine replied:
there is no sense - one have to generate js and actually add variables as code body.

xxxz commented on third_party/libadblockplus/grit/resources/elemhide_for_selector.jst:
```diff
@@ -0,0 +1,9 @@
+let elements = document.querySelectorAll(selector);
+if (elements.length > 0) {
+  for (let element of elements) {
+   if (element.src == '{{url}}')
```
Is somewhere here a possibility for JS injection?
Consider, e.g. url like `https://some/url/image' && alert("something-evil") && '' == '`.
E.g. we could take a look at https://github.com/joliss/js-string-escape/blob/master/index.js and sanitize the string in C++ before passing it to `replaceString`.
cc=@hfiguiere, @shoniko do you have some recommendations?

4ntoine replied:
i've updated the code and now escaping `filename_with_query`. there is no need to escape `url` as it comes from chromium and it's canonical.

xxxz replied:
What does "canonical" refer to here?

4ntoine replied:
i'd better use the world "valid":
https://cs.chromium.org/chromium/src/url/gurl.h?q=url.h&sq=package:chromium&g=0&l=112

so here i've  meant there is no JS injected here.

xxxz replied:
It would be better to have tests. Could you please ask testers in the issue description to test it with various malicious URLs, also give a link to this comment thread?

4ntoine replied:
sure, https://gitlab.com/eyeo/adblockplus/chromium/issues/12#note_97530896

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

 Escaping filename+query string in Javascript.

4ntoine commented on chrome/browser/android/adblock/adblock_bridge.cc:
```diff
@@ -84,6 +84,32 @@ std::string replaceString(
     return subject;
 }
 
+std::string escapeJavaScriptString(const std::string& subject) {
```
taken from https://github.com/joliss/js-string-escape/blob/master/index.js
and not taking into account utf-16 characters.

xxxz replied:
Why not to convert UTF-16? In UTF-8 they are
\\u2028 -> E2 80 A8
\\u2028 -> E2 80 A9
https://r12a.github.io/app-conversion/

Tell the truth I would just escape all non-ASCII symbols to the format "\\uXXXX". Maybe we can do it as another issue.

Additionally what about adding just in case [*backspace*, *vtab*, *CR* and *NUL*](https://github.com/OWASP/owasp-java-encoder/blob/master/core/src/test/java/org/owasp/encoder/JavaScriptEncoderTest.java).

I think we should soon start to write tests for our code somehow.

@hfiguiere any comments?

4ntoine replied:
@xxxz: i'm not aware of UTF details in C++ like special "widechar" string and all this stuff. If you feel we need it i'd ask somebody else to contribute here. BTW since in Chromium it's still `std::string` used i wonder how it works now for utf-16 characters

xxxz replied:
Just check if the sequence of characters is `0xE2 0x80 0xA8` then replace it by the string `"\\\\u2028"`. We operate only with UTF-8 strings here, the generated JS is UTF-8 string too.

4ntoine replied:
here we iterate over single (one-byte?) characters. If we want to check/replace multibyte characters we need just replace strings, not characters. So basically we need to change the whole function

xxxz replied:
yes, one can do it by using index-based loop

hfiguiere replied:
I would just check for \u2028 and \u2029 (their utf-8 form) to properly escape them and leave any other non-ASCII alone. As mentioned above this mean not using a range iterator but an index based iterator.

shoniko reviewed (approved):
LGTM

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Fixed JS for multiple blocked resources in  frame.

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Removed misleading comment (not actual anymore).

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

 Removed "function already defined" check.

4ntoine commented on third_party/libadblockplus/grit/resources/elemhide_for_selector.jst:
```diff
@@ -0,0 +1,13 @@
+// {{filename_with_query}} and {{url}} will be replaced with actual variable values
+
+function elemhideForSelector(selector) {
```
not checking if function is already defined:
1) redefinition of function does not lead to JS error (in contrast to `let` variable redefinition)
2) we already redefine `hideElement()` function, so one more does not hurt too much.

xxxz replied:
Yes and no, it's definitely sub-optimal because in this case [last declared](https://www.ecma-international.org/ecma-262/#sec-functiondeclarationinstantiation) function is used, so we have to compile and use a new function each time.

4ntoine replied:
then we need to do or not to do this for both scripts. And since the first one will come from `core` module we will have to spread this logics to core or wrap it into one more script that just checked if the function from core is already defined or define it otherwise. We're going too deeply in the details imho

xxxz replied:
Yeah, I would recommend in a future to rather create an object with the functions and check whether the object is put in the global scope or not.

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

Escaping u2028 and u2029 too.

4ntoine commented on chrome/browser/android/adblock/adblock_bridge.cc:
```diff
@@ -84,6 +84,35 @@ std::string replaceString(
+    }
+  }
+
+  return replaceString(replaceString(out,
```
As for me working with strings is more clear and readable (though slower) instead of playing with bytes and index iterator like: `if length() >= 3 && [i-2] == 0x.. && [i-1] == 0x..`.

xxxz replied:
It will not work because UTF-8 string is a sequence of bytes, you need to search exactly for bytes, e.g. `0xE2 0x80 0xA8` then replace them by a sequence of ASCII characters `\u2028`.

4ntoine replied:
why is "u2028" string not equal to `{ 0xE2, 0x80, 0xA8 }` byte array in internal representation ?

xxxz replied:
`std::string` is just an array of bytes, the sub-string "u2028" is five bytes, 'u', '2' and so on, and these bytes are not the bytes we are looking for.

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

 Replaced previous simplified version of impl used for elemhiding with elemhiding rules of escaping function with new one.

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

 Init strings as byte arrays.

xxxz commented on chrome/browser/android/adblock/adblock_bridge.cc:
```diff
@@ -84,8 +84,39 @@ std::string replaceString(
 
-std::string escapeSelector(std::string selector) {
-  return replaceString(replaceString(selector, "\\", "\\\\"), "\"", "\\\"");
+const char U2028_BYTES[] = { 0xE2, 0x80, 0xA8 };
```
does it compile? I would expect that the cast is needed here, not on the next line.

4ntoine replied:
yes, i always compile before pushing

xxxz replied:
I have just checked, it's not working, it can work because maybe chromium compiling scripts define `char` another way, could you please change `char` in the array to `uint8_t`?
```
const uint8_t U2028_BYTES[] = { 0xE2, 0x80, 0xA8 };
const std::string U2028 = reinterpret_cast<char const*>(U2028_BYTES);
```

xxxz commented on chrome/browser/android/adblock/adblock_bridge.cc:
```diff
@@ -84,8 +84,39 @@ std::string replaceString(
+  }
+
+  return replaceString(replaceString(out,
+    U2028, "\\" + U2028),
```
here the replacement should be `"\\u2028"`

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

* uint8_t

Commit:
Issue #12 - Blocked ads is not element hidden on 50statesclassifieds.com

* "\\u2028"

xxxz commented:
LGTM
